gh-127833: lexical analysis: Improve section on Names (#131474)
python/cpython: Doc/reference/lexical_analysis.rst

.. _identifiers:

Names (identifiers and keywords)
================================

.. index:: identifier, name

:data:`~token.NAME` tokens represent *identifiers*, *keywords*, and
*soft keywords*.

Within the ASCII range (U+0001..U+007F), the valid characters for names
include the uppercase and lowercase letters (``A`` through ``Z`` and ``a`` to

> ``z``), the underscore ``_`` and, except for the first character, the digits

Reviewer: ```suggestion
include the uppercase and lowercase letters, the underscore ``_`` and, except
for the first character, the digits
```

Do we really need to explain this here too?

This is just below:

> Besides ``A-Z``, ``a-z``, ``_`` and ``0-9`` ...

Author: I'd definitely keep it here; making it clear that “letters in the ASCII range” is ``A-Z`` and ``a-z``.
Perhaps shorten it:
```suggestion
include the uppercase and lowercase letters (``A-Z`` and ``a-z``),
the underscore ``_`` and, except for the first character, the digits
```

I'd be OK with deduplicating the later occurence, but something like “ASCII characters as listed above” is almost as long as repeating the lists. This isn't code, it can be a little WET.

Reviewer: That would be better :-)

* Unicode category ``<Nd>`` - decimal numbers (includes ``0`` to ``9``)
* Unicode category ``<Pc>`` - connector punctuations
* Unicode category ``<Mn>`` - nonspacing marks
* Unicode category ``<Mc>`` - spacing combining marks
* ``<Other_ID_Continue>`` - another explicit set of characters in
  `PropList.txt`_ to support backwards compatibility

Unicode categories use the version of the Unicode Character Database as
included in the :mod:`unicodedata` module.

These sets are based on the Unicode standard annex `UAX-31`_.
See also :pep:`3131` for further details.

Even more formally, names are described by the following lexical definitions:

.. grammar-snippet::
   :group: python-grammar

   NAME:         `xid_start` `xid_continue`*
   id_start:     <Lu> | <Ll> | <Lt> | <Lm> | <Lo> | <Nl> | "_" | <Other_ID_Start>
   id_continue:  `id_start` | <Nd> | <Pc> | <Mn> | <Mc> | <Other_ID_Continue>
   xid_start:    <all characters in `id_start` whose NFKC normalization is
                  in (`id_start` `xid_continue`*)">
   xid_continue: <all characters in `id_continue` whose NFKC normalization is
                  in (`id_continue`*)">
   identifier:   <`NAME`, except keywords>

> A non-normative HTML file listing all valid identifier characters for Unicode

Reviewer: ```suggestion
A non-normative text file listing all valid identifier characters for Unicode
```

It refers to a text file.

> 16.0.0 can be found at

Reviewer: ```suggestion
15.1.0 can be found at
```
> Unicode categories use the version of the Unicode Character Database as
included in the :mod:`unicodedata` module.

According to the [docs](https://docs.python.org/3/library/unicodedata.html#module-unicodedata) that is 15.1.0 at the moment, maybe we should link to that one instead?

Author: [`main` is at 16.0](https://docs.python.org/3.14/library/unicodedata.html#module-unicodedata).

Reviewer: Oh apologies

> https://www.unicode.org/Public/16.0.0/ucd/DerivedCoreProperties.txt

Reviewer: Maybe the link could be moved below too? Inline links are nice ;-)
